_search-plugins/search-pipelines/system-generated-search-processors.md

```diff
@@ -0,0 +1,60 @@
+
+Example:
+```json
+PUT _cluster/settings
+{
+  "persistent": {
+    "cluster.search.enabled_system_generated_factories": [
+      "mmr_over_sample_factory",
+      "mmr_rerank_factory"
+    ]
+  }
+}
+```
+{% include copy-curl.html %}
+
+
+
+Search processors can be of the following types:
+
+- [System generated search request processors](#system-generated-search-request-processors)
+- [System generated search response processors](#system-generated-search-response-processors)
+- [System generated search phase results processors](#system-generated-search-phase-results-processors)
+
+# System generated search request processors
+
+| Processor name    | Processor factory name    | Execution stage     | Trigger condition                                           | Description                                                                                                                                                                |
+|-------------------|---------------------------| ------------------- |-------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `mmr_over_sample` | `mmr_over_sample_factory` | `POST_USER_DEFINED` | Triggered when a search request includes the mmr extension  | Modifies the query size and `k` value of the k-NN query to oversample candidates for MMR re-ranking. This processor runs after any user-defined search request processors. |
+
+The execution stage determines whether a system-generated processor runs before or after user-defined processors of the same type.
+
+# System generated search response processors
+
+| Processor name | Processor factory name | Execution stage     | Trigger condition                                          | Description                                                                                                                                                     |
+|----------------|------------------------|---------------------|------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `mmr_rerank`   | `mmr_rerank_factory`   | `PRE_USER_DEFINED`  | Triggered when a search request includes the mmr extension | Re-ranks the oversampled results using MMR and reduces them to the original query size. This processor runs before any user-defined search response processors. |
+
+The execution stage determines whether a system-generated processor runs before or after user-defined processors of the same type.
+
+# System generated search phase results processors
+
+We don't have any for now.
+
+# Limitation
+
+## One processor per type and execution stage
```

Comment:
@bzhangam @heemin32 Given that there are 2 processor types (request and response) and 2 execution stages (PRE_USER_DEFINED and POST_USER_DEFINED), does this mean that a single search request can have up to 4 system-generated processors (request processor at PRE_USER_DEFINED stage, request processor at POST_USER_DEFINED stage, response processor at PRE_USER_DEFINED stage, and response processor at POST_USER_DEFINED stage)? Or is the limit actually 2 total processors (one request + one response) regardless of execution stage?

Reply:
Total system generated processors which can be run is 6. 
Three type: SearchRequestProcessor, SearchPhaseResultProcessor, SearchResponseProcessor
Two stage: PRE_USER_DEFIEND, POST_USER_DEFINED

https://github.com/opensearch-project/OpenSearch/issues/19062#issuecomment-3202355704